File: kotlin-desktop-toolkit/src/main/kotlin/org/jetbrains/desktop/Geometry.kt
package org.jetbrains.desktop

typealias PhysicalPixels = Double
typealias LogicalPixels = Double

data class PhysicalSize(val width: PhysicalPixels, val height: PhysicalPixels) {
    companion object

    fun toLogical(scale: Double) = LogicalSize(width / scale, height / scale)
}

data class PhysicalPoint(val x: PhysicalPixels, val y: PhysicalPixels) {
    companion object

    fun toLogical(scale: Double) = LogicalPoint(x / scale, y / scale)
}

data class LogicalSize(val width: LogicalPixels, val height: LogicalPixels) {
    companion object

    fun toPhysical(scale: Double) = PhysicalSize(width * scale, height * scale)
}

data class LogicalPoint(val x: LogicalPixels, val y: LogicalPixels) {
    companion object {
        val Zero = LogicalPoint(0.0, 0.0)
    }

    fun toPhysical(scale: Double) = PhysicalPoint(x * scale, y * scale)
}

File: native/desktop-macos/src/geometry.rs
pub type PhysicalPixels = f64;
pub type LogicalPixels = f64;

#[repr(C)]
#[derive(Debug, Clone, Copy)]
pub struct PhysicalSize {
    pub width: PhysicalPixels,
    pub height: PhysicalPixels,
}

#[repr(C)]
#[derive(Debug, Clone, Copy)]
pub struct PhysicalPoint {
    pub x: PhysicalPixels,
    pub y: PhysicalPixels,
}

#[repr(C)]
#[derive(Debug, Clone, Copy, Default)]
pub struct LogicalSize {
    pub width: LogicalPixels,
    pub height: LogicalPixels,
}

#[repr(C)]
#[derive(Debug, Clone, Copy, Default)]
pub struct LogicalPoint {
    pub x: LogicalPixels,
    pub y: LogicalPixels,
}

#[repr(C)]
#[derive(Debug, Default)]
pub struct LogicalRect {
    // the point closest to coordinates origin
    pub origin: LogicalPoint,
    pub size: LogicalSize,
}

impl LogicalRect {
    pub const fn new(origin: LogicalPoint, size: LogicalSize) -> Self {
        Self { origin, size }
    }
}

#[repr(C)]
#[derive(Debug, Clone, Copy)]
pub struct Color {
    pub red: f64,
    pub green: f64,
    pub blue: f64,
    pub alpha: f64,
}
